File: pythonlib/winfox/rdp/locator.py
import asyncio
import json
import time
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from camoufox._rdp_legacy_impl import RDPPage


class _Locator:
    """Playwright-compatible locator for RDPPage."""

    def __init__(
        self,
        page: "RDPPage",
        selector: str,
        index: Optional[int] = None,
        mode: str = "css",
        value: Optional[str] = None,
        exact: bool = False,
        role_name: Optional[str] = None,
        role_name_exact: bool = False,
        has_text: Optional[str] = None,
        has_text_exact: bool = False,
        parent: Optional["_Locator"] = None,
    ):
        self._page = page
        self._selector = selector
        self._index = index
        self._mode = mode
        self._value = value
        self._exact = exact
        self._role_name = role_name
        self._role_name_exact = role_name_exact
        self._has_text = has_text
        self._has_text_exact = has_text_exact
        self._parent = parent

    def _collection_expr(self, root: str = "document") -> str:
        if self._parent is not None:
            root = self._parent._to_css_and_js()
        root_base = f"(({root}).body||({root}).documentElement||({root}))"
        if self._mode == "text":
            text = (self._value or "").replace("\\", "\\\\").replace("'", "\\'")
            matcher = f"txt.trim() === '{text}'" if self._exact else f"txt.includes('{text}')"
            return (
                f"(function(){{ var scope={root}; var out=[]; var seen=new Set(); var tw=document.createTreeWalker({root_base}, NodeFilter.SHOW_TEXT);"
                f"while(tw.nextNode()){{ var txt=tw.currentNode.textContent||''; var el=tw.currentNode.parentElement; if(el && ({matcher}) && !seen.has(el)){{ seen.add(el); out.push(el); }} }} return out; }})()"
            )

        if self._mode == "placeholder":
            value = (self._value or "").replace("\\", "\\\\").replace("'", "\\'")
            matcher = f"p.trim() === '{value}'" if self._exact else f"p.includes('{value}')"
            return f"Array.from(({root}).querySelectorAll('input[placeholder], textarea[placeholder]')).filter(el => {{ var p = el.getAttribute('placeholder') || ''; return {matcher}; }})"

        if self._mode == "label":
            value = (self._value or "").replace("\\", "\\\\").replace("'", "\\'")
            matcher = f"txt.trim() === '{value}'" if self._exact else f"txt.includes('{value}')"
            return f"(function(){{ var scope={root}; return Array.from(scope.querySelectorAll('label')).map(label => {{ var txt=(label.innerText||label.textContent||''); if (!({matcher})) return null; var target=null; var htmlFor=label.getAttribute('for'); if (htmlFor && scope.getElementById) target=scope.getElementById(htmlFor); if(!target) target=label.querySelector('input, textarea, select, button'); return target; }}).filter(Boolean); }})()"

        if self._mode == "role":
            role = (self._value or "").replace("\\", "\\\\").replace("'", "\\'")
            role_name = (self._role_name or "").replace("\\", "\\\\").replace("'", "\\'")
            name_match = "true" if self._role_name is None else (f"accName.trim() === '{role_name}'" if self._role_name_exact else f"accName.includes('{role_name}')")
            return f"""(function(){{
              const scope = {root};
              const all = Array.from(scope.querySelectorAll('*'));
              function implicitRole(el) {{
                const tag = (el.tagName || '').toLowerCase();
                const type = (el.getAttribute('type') || '').toLowerCase();
                if (tag === 'button') return 'button';
                if (tag === 'a' && el.hasAttribute('href')) return 'link';
                if (tag === 'textarea') return 'textbox';
                if (tag === 'input' && ['text','email','search','url','tel','password'].includes(type)) return 'textbox';
                if (tag === 'input' && type === 'checkbox') return 'checkbox';
                if (tag === 'input' && type === 'radio') return 'radio';
                if (tag === 'select') return 'combobox';
                if (tag === 'input' && ['button','submit','reset'].includes(type)) return 'button';
                return null;
              }}
              function accessibleName(el) {{
                const ariaLabel = el.getAttribute('aria-label');
                if (ariaLabel) return ariaLabel.trim();
                const labelledBy = el.getAttribute('aria-labelledby');
                if (labelledBy) return labelledBy.split(/\s+/).map(id => scope.getElementById ? scope.getElementById(id) : null).filter(Boolean).map(n => (n.innerText || n.textContent || '').trim()).join(' ').trim();
                if (el.labels && el.labels.length) return Array.from(el.labels).map(l => (l.innerText || l.textContent || '').trim()).join(' ').trim();
                const tag = (el.tagName || '').toLowerCase();
                const type = (el.getAttribute('type') || '').toLowerCase();
                if (tag === 'input' && ['button','submit','reset'].includes(type)) return (el.value || '').trim();
                return ((el.innerText || el.textContent || '') || '').trim();
              }}
              return all.filter(el => {{
                const explicitRole = (el.getAttribute('role') || '').trim();
                const actualRole = explicitRole || implicitRole(el);
                if (actualRole !== '{role}') return false;
                const accName = accessibleName(el);
                return {name_match};
              }});
            }})()"""

        if self._mode == "testid":
            value = (self._value or "").replace("\\", "\\\\").replace("'", "\\'")
            return f"Array.from(({root}).querySelectorAll('[data-testid]')).filter(el => (el.getAttribute('data-testid') || '') === '{value}')"

        sel = self._selector
        if sel.startswith("text="):
            text = sel[5:]
            text = text.replace("\\", "\\\\").replace("'", "\\'")
            return (
                f"(function(){{ var out=[]; var seen=new Set(); var tw=document.createTreeWalker({root_base}, NodeFilter.SHOW_TEXT);"
                f"while(tw.nextNode()){{ var txt=tw.currentNode.textContent||''; var el=tw.currentNode.parentElement; if(el && txt.includes('{text}') && !seen.has(el)){{ seen.add(el); out.push(el); }} }} return out; }})()"
            )
        if sel.startswith("css:") or sel.startswith("css="):
            sel = sel[4:]
        sel_escaped = sel.replace(chr(39), chr(92) + chr(39))
        return f"Array.from(({root}).querySelectorAll('{sel_escaped}'))"

    def _apply_text_filter(self, collection_expr: str) -> str:
        if self._has_text is None:
            return collection_expr
        text = self._has_text.replace("\\", "\\\\").replace("'", "\\'")
        matcher = f"txt.trim() === '{text}'" if self._has_text_exact else f"txt.includes('{text}')"
        return f"(function(){{ return ({collection_expr}).filter(el => {{ var txt=(el.innerText||el.textContent||''); return {matcher}; }}); }})()"

    def _to_css_and_js(self) -> str:
        collection_expr = self._apply_text_filter(self._collection_expr("document"))
        if self._index is None:
            return f"(({collection_expr})[0] || null)"
        if self._index == -1:
            return f"(function(){{ var els=({collection_expr}); return els.length ? els[els.length-1] : null; }})()"
        return f"(function(){{ var els=({collection_expr}); return els.length>{self._index} ? els[{self._index}] : null; }})()"

    def first(self) -> "_Locator":
        return _Locator(self._page, self._selector, index=0, mode=self._mode, value=self._value, exact=self._exact, role_name=self._role_name, role_name_exact=self._role_name_exact, has_text=self._has_text, has_text_exact=self._has_text_exact, parent=self._parent)

    def nth(self, index: int) -> "_Locator":
        return _Locator(self._page, self._selector, index=index, mode=self._mode, value=self._value, exact=self._exact, role_name=self._role_name, role_name_exact=self._role_name_exact, has_text=self._has_text, has_text_exact=self._has_text_exact, parent=self._parent)

    def last(self) -> "_Locator":
        return _Locator(self._page, self._selector, index=-1, mode=self._mode, value=self._value, exact=self._exact, role_name=self._role_name, role_name_exact=self._role_name_exact, has_text=self._has_text, has_text_exact=self._has_text_exact, parent=self._parent)

    def filter(self, has_text: Optional[str] = None, exact: bool = False) -> "_Locator":
        return _Locator(self._page, self._selector, index=self._index, mode=self._mode, value=self._value, exact=self._exact, role_name=self._role_name, role_name_exact=self._role_name_exact, has_text=has_text, has_text_exact=exact, parent=self._parent)

    def locator(self, selector: str) -> "_Locator":
        return _Locator(self._page, selector, parent=self)

    async def wait_for(self, state: str = "visible", timeout: int = 5000) -> None:
        find_js = self._to_css_and_js()
        wfs_key = f"_l{int(time.time() * 1000) % 100000}"

        if state == "hidden":
            setup_js = (
                f"(function(){{"
                f"  if(!window._ws)window._ws={{}};"
                f"  if (({find_js}) === null) {{ window._ws['{wfs_key}']='ok'; return; }}"
                f"  var obs = new MutationObserver(function(){{"
                f"    if (({find_js}) === null) {{ obs.disconnect(); window._ws['{wfs_key}']='ok'; }}"
                f"  }});"
                f"  obs.observe(document.body||document.documentElement,"
                f"    {{childList:true,subtree:true,attributes:true}});"
                f"  setTimeout(function(){{ obs.disconnect(); if(!window._ws['{wfs_key}']) window._ws['{wfs_key}']='timeout'; }},{timeout});"
                f"}})()"
            )
        else:
            vis = (
                "if(r.width===0&&r.height===0) return null; "
                if state == "visible"
                else ""
            )
            setup_js = (
                f"(function(){{"
                f"  if(!window._ws)window._ws={{}};"
                f"  function chk(){{"
                f"    var el={find_js}; if(!el) return null;"
                f"    var r=el.getBoundingClientRect(); {vis}"
                f"    return JSON.stringify({{x:r.x,y:r.y,w:r.width,h:r.height}});"
                f"  }}"
                f"  var hit=chk(); if(hit){{ window._ws['{wfs_key}']=hit; return; }}"
                f"  var obs=new MutationObserver(function(){{"
                f"    var hit=chk(); if(hit){{ obs.disconnect(); window._ws['{wfs_key}']=hit; }}"
                f"  }});"
                f"  obs.observe(document.body||document.documentElement,"
                f"    {{childList:true,subtree:true,attributes:true}});"
                f"  setTimeout(function(){{ obs.disconnect(); if(!window._ws['{wfs_key}']) window._ws['{wfs_key}']='timeout'; }},{timeout});"
                f"}})()"
            )

        await self._page.evaluate(setup_js)

        deadline = time.time() + (timeout / 1000)
        while time.time() < deadline:
            val = await self._page.evaluate(f"(window._ws||{{}})['{wfs_key}']")
            if val and val != "null":
                await self._page.evaluate(f"try{{delete window._ws['{wfs_key}']}}catch(e){{}}")
                if val == "timeout":
                    raise TimeoutError(f"Locator '{self._selector}' not {state} within {timeout}ms")
                return
            await asyncio.sleep(0.1)
        try:
            await self._page.evaluate(f"try{{delete window._ws['{wfs_key}']}}catch(e){{}}")
        except Exception:
            pass
        raise TimeoutError(f"Locator '{self._selector}' not {state} within {timeout}ms")

    async def click(self, timeout: int = 5000) -> None:
        find_js = self._to_css_and_js()
        deadline = time.time() + (timeout / 1000)
        while time.time() < deadline:
            try:
                js = (
                    f"(function(){{ var el = {find_js}; if(!el) return null; "
                    f"var r = el.getBoundingClientRect(); "
                    f"if(r.width===0&&r.height===0) return null; "
                    f"return JSON.stringify({{x:r.x+r.width/2,y:r.y+r.height/2,w:r.width}}); }})()"
                )
                result = await self._page.evaluate(js)
                if result and isinstance(result, str):
                    pos = json.loads(result)
                    await self._page.mouse.click_smooth(pos["x"], pos["y"], target_width=pos.get("w", 50))
                    return
            except Exception:
                pass
            await asyncio.sleep(0.3)
        raise TimeoutError(f"Locator '{self._selector}' not clickable within {timeout}ms")

    async def text_content(self) -> Optional[str]:
        find_js = self._to_css_and_js()
        result = await self._page.evaluate(f"(function(){{ var el = {find_js}; return el ? el.textContent : null; }})()")
        return result

    async def inner_text(self) -> Optional[str]:
        find_js = self._to_css_and_js()
        result = await self._page.evaluate(f"(function(){{ var el = {find_js}; return el ? el.innerText : null; }})()")
        return result

    async def get_attribute(self, name: str) -> Optional[str]:
        find_js = self._to_css_and_js()
        name_escaped = name.replace("'", "\\'")
        result = await self._page.evaluate(f"(function(){{ var el = {find_js}; return el ? el.getAttribute('{name_escaped}') : null; }})()")
        return result

    async def count(self) -> int:
        collection_expr = self._apply_text_filter(self._collection_expr("document"))
        result = await self._page.evaluate(f"({collection_expr}).length")
        return result or 0

    async def exists(self) -> bool:
        return (await self.count()) > 0

    async def is_visible(self) -> bool:
        find_js = self._to_css_and_js()
        result = await self._page.evaluate(
            f"""(function(){{
                var el = {find_js};
                if (!el) return false;
                var style = window.getComputedStyle(el);
                if (!style) return false;
                if (style.display === 'none' || style.visibility === 'hidden' || style.opacity === '0') return false;
                var r = el.getBoundingClientRect();
                return !!(r.width > 0 && r.height > 0);
            }})()"""
        )
        return bool(result)

    async def is_hidden(self) -> bool:
        return not await self.is_visible()

File: packages/winfox/src/rdp/locator.ts
import type { RDPPage } from "./page.js";

export class Locator {
  private _page: RDPPage;
  private _selector: string;
  private _index: number | null;
  private _mode: string | null;
  private _value: string | null;
  private _exact?: boolean;
  private _roleName?: string | null;
  private _roleNameExact?: boolean;
  private _hasText?: string | null;
  private _hasTextExact?: boolean;
  private _parent: Locator | null;

  constructor(page: RDPPage, selector: string, options: {
    index?: number | null, mode?: string | null, value?: string | null, exact?: boolean | undefined,
    roleName?: string | null | undefined, roleNameExact?: boolean | undefined, hasText?: string | null | undefined, hasTextExact?: boolean | undefined, parent?: Locator | null
  } = {}) {
    this._page = page;
    this._selector = selector;
    this._index = options.index ?? null;
    this._mode = options.mode ?? null;
    this._value = options.value ?? null;
    this._exact = options.exact ?? false;
    this._roleName = options.roleName ?? null;
    this._roleNameExact = options.roleNameExact ?? false;
    this._hasText = options.hasText ?? null;
    this._hasTextExact = options.hasTextExact ?? false;
    this._parent = options.parent ?? null;
  }

  private _collectionExpr(root: string = "document"): string {
    if (this._parent !== null) {
      root = this._parent._elementExpr();
    }
    const rootBase = `((${root}).body||(${root}).documentElement||(${root}))`;

    if (this._mode === "text") {
      const text = (this._value || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      const matcher = this._exact ? `txt.trim() === '${text}'` : `txt.includes('${text}')`;
      return `(function(){ var out=[]; var seen=new Set(); var tw=document.createTreeWalker(${rootBase}, NodeFilter.SHOW_TEXT); while(tw.nextNode()){ var txt=tw.currentNode.textContent||''; var el=tw.currentNode.parentElement; if(el && (${matcher}) && !seen.has(el)){ seen.add(el); out.push(el); } } return out; })()`;
    }
    if (this._mode === "placeholder") {
      const value = (this._value || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      const matcher = this._exact ? `p.trim() === '${value}'` : `p.includes('${value}')`;
      return `Array.from((${root}).querySelectorAll('input[placeholder], textarea[placeholder]')).filter(el => { var p = el.getAttribute('placeholder') || ''; return ${matcher}; })`;
    }
    if (this._mode === "label") {
      const value = (this._value || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      const matcher = this._exact ? `txt.trim() === '${value}'` : `txt.includes('${value}')`;
      return `(function(){ var scope=${root}; return Array.from(scope.querySelectorAll('label')).map(label => { var txt=(label.innerText||label.textContent||''); if (!(${matcher})) return null; var target=null; var htmlFor=label.getAttribute('for'); if (htmlFor && scope.getElementById) target=scope.getElementById(htmlFor); if(!target) target=label.querySelector('input, textarea, select, button'); return target; }).filter(Boolean); })()`;
    }
    if (this._mode === "role") {
      const role = (this._value || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      const roleName = (this._roleName || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      const nameMatch = this._roleName === null ? "true" : (this._roleNameExact ? `accName.trim() === '${roleName}'` : `accName.includes('${roleName}')`);
      return `(function(){
        const scope = ${root};
        const all = Array.from(scope.querySelectorAll('*'));
        function implicitRole(el) {
          const tag = (el.tagName || '').toLowerCase();
          const type = (el.getAttribute('type') || '').toLowerCase();
          if (tag === 'button') return 'button';
          if (tag === 'a' && el.hasAttribute('href')) return 'link';
          if (tag === 'textarea') return 'textbox';
          if (tag === 'input' && ['text','email','search','url','tel','password'].includes(type)) return 'textbox';
          if (tag === 'input' && type === 'checkbox') return 'checkbox';
          if (tag === 'input' && type === 'radio') return 'radio';
          if (tag === 'select') return 'combobox';
          if (tag === 'input' && ['button','submit','reset'].includes(type)) return 'button';
          return null;
        }
        function accessibleName(el) {
          const ariaLabel = el.getAttribute('aria-label');
          if (ariaLabel) return ariaLabel.trim();
          const labelledBy = el.getAttribute('aria-labelledby');
          if (labelledBy) return labelledBy.split(/\\s+/).map(id => scope.getElementById ? scope.getElementById(id) : null).filter(Boolean).map(n => (n.innerText || n.textContent || '').trim()).join(' ').trim();
          if (el.labels && el.labels.length) return Array.from(el.labels).map(l => (l.innerText || l.textContent || '').trim()).join(' ').trim();
          const tag = (el.tagName || '').toLowerCase();
          const type = (el.getAttribute('type') || '').toLowerCase();
          if (tag === 'input' && ['button','submit','reset'].includes(type)) return (el.value || '').trim();
          return ((el.innerText || el.textContent || '') || '').trim();
        }
        return all.filter(el => {
          const explicitRole = (el.getAttribute('role') || '').trim();
          const actualRole = explicitRole || implicitRole(el);
          if (actualRole !== '${role}') return false;
          const accName = accessibleName(el);
          return ${nameMatch};
        });
      })()`;
    }
    if (this._mode === "testid") {
      const value = (this._value || "").replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      return `Array.from((${root}).querySelectorAll('[data-testid]')).filter(el => (el.getAttribute('data-testid') || '') === '${value}')`;
    }

    let sel = this._selector;
    if (sel.startsWith("text=")) {
      const text = sel.substring(5).replace(/\\/g, "\\\\").replace(/'/g, "\\'");
      return `(function(){ var out=[]; var seen=new Set(); var tw=document.createTreeWalker(${rootBase}, NodeFilter.SHOW_TEXT); while(tw.nextNode()){ var txt=tw.currentNode.textContent||''; var el=tw.currentNode.parentElement; if(el && txt.includes('${text}') && !seen.has(el)){ seen.add(el); out.push(el); } } return out; })()`;
    }
    if (sel.startsWith("css:") || sel.startsWith("css=")) {
      sel = sel.substring(4);
    }
    const selEscaped = sel.replace(/'/g, "\\'");
    return `Array.from((${root}).querySelectorAll('${selEscaped}'))`;
  }

  private _applyTextFilter(collectionExpr: string): string {
    if (this._hasText === null) return collectionExpr;
    const text = this._hasText!.replace(/\\/g, "\\\\").replace(/'/g, "\\'");
    const matcher = this._hasTextExact ? `txt.trim() === '${text}'` : `txt.includes('${text}')`;
    return `(function(){ return (${collectionExpr}).filter(el => { var txt=(el.innerText||el.textContent||''); return ${matcher}; }); })()`;
  }

  private _toCssAndJs(): string {
    const collectionExpr = this._applyTextFilter(this._collectionExpr("document"));
    if (this._index === null) return `((${collectionExpr})[0] || null)`;
    if (this._index === -1) return `(function(){ var els=(${collectionExpr}); return els.length ? els[els.length-1] : null; })()`;
    return `(function(){ var els=(${collectionExpr}); return els.length>${this._index} ? els[${this._index}] : null; })()`;
  }

  public _elementExpr(): string {
    return this._toCssAndJs();
  }

  first(): Locator {
    return new Locator(this._page, this._selector, { index: 0, mode: this._mode, value: this._value, exact: this._exact, roleName: this._roleName, roleNameExact: this._roleNameExact, hasText: this._hasText, hasTextExact: this._hasTextExact, parent: this._parent });
  }

  nth(index: number): Locator {
    return new Locator(this._page, this._selector, { index, mode: this._mode, value: this._value, exact: this._exact, roleName: this._roleName, roleNameExact: this._roleNameExact, hasText: this._hasText, hasTextExact: this._hasTextExact, parent: this._parent });
  }

  last(): Locator {
    return new Locator(this._page, this._selector, { index: -1, mode: this._mode, value: this._value, exact: this._exact, roleName: this._roleName, roleNameExact: this._roleNameExact, hasText: this._hasText, hasTextExact: this._hasTextExact, parent: this._parent });
  }

  filter(options: { hasText?: string, exact?: boolean }): Locator {
    return new Locator(this._page, this._selector, { index: this._index, mode: this._mode, value: this._value, exact: this._exact, roleName: this._roleName, roleNameExact: this._roleNameExact, hasText: options.hasText, hasTextExact: options.exact, parent: this._parent });
  }

  locator(selector: string): Locator {
    return new Locator(this._page, selector, { parent: this });
  }

  async waitFor(options: { state?: "visible" | "hidden", timeout?: number } = {}): Promise<void> {
    const state = options.state || "visible";
    const timeout = options.timeout || 5000;
    const findJs = this._toCssAndJs();
    const wfsKey = `_l${(Date.now() % 100000)}`;

    let setupJs = "";
    if (state === "hidden") {
      setupJs = `(function(){
        if(!window._ws)window._ws={};
        if ((${findJs}) === null) { window._ws['${wfsKey}']='ok'; return; }
        var obs = new MutationObserver(function(){
          if ((${findJs}) === null) { obs.disconnect(); window._ws['${wfsKey}']='ok'; }
        });
        obs.observe(document.body||document.documentElement, {childList:true,subtree:true,attributes:true});
        setTimeout(function(){ obs.disconnect(); if(!window._ws['${wfsKey}']) window._ws['${wfsKey}']='timeout'; },${timeout});
      })()`;
    } else {
      const vis = state === "visible" ? "if(r.width===0&&r.height===0) return null;" : "";
      setupJs = `(function(){
        if(!window._ws)window._ws={};
        function chk(){
          var el=${findJs}; if(!el) return null;
          var r=el.getBoundingClientRect(); ${vis}
          return JSON.stringify({x:r.x,y:r.y,w:r.width,h:r.height});
        }
        var hit=chk(); if(hit){ window._ws['${wfsKey}']=hit; return; }
        var obs=new MutationObserver(function(){
          var hit=chk(); if(hit){ obs.disconnect(); window._ws['${wfsKey}']=hit; }
        });
        obs.observe(document.body||document.documentElement, {childList:true,subtree:true,attributes:true});
        setTimeout(function(){ obs.disconnect(); if(!window._ws['${wfsKey}']) window._ws['${wfsKey}']='timeout'; },${timeout});
      })()`;
    }

    await this._page.evaluate(setupJs);
    const deadline = Date.now() + timeout;
    while (Date.now() < deadline) {
      const val = await this._page.evaluate(`(window._ws||{})['${wfsKey}']`);
      if (val && val !== "null") {
        await this._page.evaluate(`try{delete window._ws['${wfsKey}']}catch(e){}`);
        if (val === "timeout") throw new Error(`Locator '${this._selector}' not ${state} within ${timeout}ms`);
        return;
      }
      await new Promise(r => setTimeout(r, 100));
    }
    try { await this._page.evaluate(`try{delete window._ws['${wfsKey}']}catch(e){}`); } catch(e) {}
    throw new Error(`Locator '${this._selector}' not ${state} within ${timeout}ms`);
  }

  async click(options: { timeout?: number } = {}): Promise<void> {
    const timeout = options.timeout || 5000;
    const findJs = this._toCssAndJs();
    const deadline = Date.now() + timeout;
    while (Date.now() < deadline) {
      try {
        const js = `(function(){ var el = ${findJs}; if(!el) return null; var r = el.getBoundingClientRect(); if(r.width===0&&r.height===0) return null; return JSON.stringify({x:r.x+r.width/2,y:r.y+r.height/2,w:r.width}); })()`;
        const result = await this._page.evaluate(js);
        if (result && typeof result === "string") {
          const pos = JSON.parse(result);
          await this._page.mouse.clickSmooth(pos.x, pos.y, pos.w || 50);
          return;
        }
      } catch (e) {}
      await new Promise(r => setTimeout(r, 300));
    }
    throw new Error(`Locator '${this._selector}' not clickable within ${timeout}ms`);
  }

  async textContent(): Promise<string | null> {
    const findJs = this._toCssAndJs();
    return await this._page.evaluate(`(function(){ var el = ${findJs}; return el ? el.textContent : null; })()`);
  }

  async innerText(): Promise<string | null> {
    const findJs = this._toCssAndJs();
    return await this._page.evaluate(`(function(){ var el = ${findJs}; return el ? el.innerText : null; })()`);
  }

  async getAttribute(name: string): Promise<string | null> {
    const findJs = this._toCssAndJs();
    const nameEscaped = name.replace(/'/g, "\\'");
    return await this._page.evaluate(`(function(){ var el = ${findJs}; return el ? el.getAttribute('${nameEscaped}') : null; })()`);
  }

  async count(): Promise<number> {
    const collectionExpr = this._applyTextFilter(this._collectionExpr("document"));
    const result = await this._page.evaluate(`(${collectionExpr}).length`);
    return result || 0;
  }

  async exists(): Promise<boolean> {
    return (await this.count()) > 0;
  }

  async isVisible(): Promise<boolean> {
    const findJs = this._toCssAndJs();
    const result = await this._page.evaluate(`(function(){
      var el = ${findJs};
      if (!el) return false;
      var style = window.getComputedStyle(el);
      if (!style) return false;
      if (style.display === 'none' || style.visibility === 'hidden' || style.opacity === '0') return false;
      var r = el.getBoundingClientRect();
      return !!(r.width > 0 && r.height > 0);
    })()`);
    return !!result;
  }

  async isHidden(): Promise<boolean> {
    return !(await this.isVisible());
  }
}
